src/Provider/AbstractProvider.php

```diff
@@ -295,22 +286,30 @@ protected function parseResponse($response)
         switch ($this->responseType) {
             case 'json':
                 $result = json_decode($response, true);
-
                 if (JSON_ERROR_NONE !== json_last_error()) {
                     throw new UnexpectedValueException('Unable to parse client response');
                 }
-
                 break;
             case 'string':
                 parse_str($response, $result);
-
                 break;
         }
 
+        $this->checkResponse($result);
+
         return $result;
     }
 
     /**
+     * Check a provider response for errors.
+     *
+     * @throws IdentityProviderException
+     * @param  array $response
+     * @return void
+     */
+    abstract protected function checkResponse(array $response);
```

Comment:
I don't see any reason for this to be part of the `ProviderInterface`, because it is a self-referencing method... it doesn't make sense to use it externally.